model/riscv_step.sail

```diff
@@ -15,36 +15,74 @@ union Step = {
   Step_Ext_Fetch_Failure  : ext_fetch_addr_error,
   Step_Fetch_Failure      : (virtaddr, ExceptionType),
   Step_Execute            : (ExecutionResult, instbits),
-  Step_Waiting            : unit,
+  Step_Waiting            : WaitReason,
 }
 
-function run_hart_waiting(step_no : nat, exit_wait : bool, instbits : instbits) -> Step = {
-  if shouldWakeForInterrupt() then {
-    if   get_config_print_instr()
-    then print_instr("interrupt exit from WAIT state at PC " ^ BitStr(PC));
-
-    hart_state = HART_ACTIVE();
-    // The waiting instruction retires successfully.  The
-    // pending interrupts will be handled in the next step.
-    Step_Execute(Retire_Success(), instbits)
-  } else if exit_wait then {
-    // There are no pending interrupts; transition out of the Wait
-    // as instructed.
-    if   get_config_print_instr()
-    then print_instr("forced exit from WAIT state at PC " ^ BitStr(PC));
-
-    hart_state = HART_ACTIVE();
-    // "When TW=1, then if WFI is executed in any
-    // less-privileged mode, and it does not complete within an
-    // implementation-specific, bounded time limit, the WFI
-    // instruction causes an illegal-instruction exception."
-    if   (cur_privilege == Machine | mstatus[TW] == 0b0)
-    then Step_Execute(Retire_Success(), instbits)
-    else Step_Execute(Illegal_Instruction(), instbits)
-  } else {
-    if   get_config_print_instr()
-    then print_instr("remaining in WAIT state at PC " ^ BitStr(PC));
-    Step_Waiting()
+function run_hart_waiting(step_no : int, wr: WaitReason, instbits : instbits, exit_wait : bool) -> Step = {
+  match (shouldWakeForInterrupt(), wr, valid_reservation(), exit_wait) {
```

Comment:
IMO this style of match is much more confusing than `if`/`else`.

```
match (boolean, .....) {
   (true, _, _, _) => then,
   ...
}
```

is literally an `if then` but ... odder. I guess we can refactor it later if you like.

Reply:
Actually, that portion will change for Sdext anyway, so we could refactor it then.

Reply:
I went ahead and refactored it, should make Sdext easier.